### ISSUE TITLE
Why JAX shows different training pattern than Pytorch?

### ISSUE THREAD

**Issue by hudsonchen** [deepmind/dm-haiku#120]
Hi all,

I noticed that Jax used together with dm-haiku shows different training dynamics than PyTorch, when using the same architecture, optimizer, and hyperparameters, initialization scheme, seeds, dataloaders, etc. Specifically,  Jax appears to show faster convergence than PyTorch and has (comparably) higher accuracy after 100 epochs. The difference seems fairly significant and and seems systematic across datasets and training runs.

You can see the described behavior in the following two minimal implementations:
[JAX](https://colab.research.google.com/drive/1opGcSFR8ockEQiFc7wnq4Af07Hd5HUPH?usp=sharing)
[PyTorch](https://colab.research.google.com/drive/1MUFZngVTkd2tli_bh38ezfVFNiRtx6IP?usp=sharing)

In the implementations above, the following factors were considered to ensure the comparison is accurate:

1. Both use a same custom CNN (a small network tailored to CIFAR-10 classification).
2. Both use Adam optimizer with learning rate 1e-3 (without learning rate decay), a min-batch size of 200 and weight decay of 5e-4.
3. Both use the CIFAR-10 data loader in torchvision (with the same transformations).
4. Both use a self-implemented cross-entropy loss function (instead of nn.functional.cross_entropy in the case of PyTorch).
5. Both use an explicit weight decay term in the loss function (instead of using the weight decay argument in the PyTorch Adam optimizer).
6. The Jax weight initialization is changed to he_uniform (which is what I understand is used in PyTorch).

I also took a closer look into the Jax and PyTorch repos and found that:

1. Both Jax and PyTorch use the same way to implement Adam.
2. Both Jax and PyTorch use float32 as the default data type.

After accounting for all of the above, I’m really at a loss what could be causing the fairly different training dynamics, especially since it makes it difficult to reproduce results for models originally implemented in PyTorch.

I would really appreciate if you could let me know if I missed something, and/or if you have an idea what could be causing the different training dynamics.

Thank you!

**Comment by tomhennigan** [comment]
Hi @hudsonchen it sounds like you're doing all of the right things to explore this. I propose we keep the discussion in google/jax#6325.
